src/test/java/io/pinecone/PineconeConfigTest.java

```diff
@@ -0,0 +1,66 @@
+package io.pinecone;
+
+import io.pinecone.configs.PineconeConfig;
+import io.pinecone.exceptions.PineconeConfigurationException;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class PineconeConfigTest {
+
+    @Test
+    public void testValidateWithNullApiKey() {
+        try {
+            new PineconeConfig(null);
+        } catch (PineconeConfigurationException expected) {
+            assertEquals(expected.getLocalizedMessage(), "The API key is required and must not be empty or null");
+        }
+    }
+    @Test
+    public void testValidateWithEmptyApiKey() {
+        try {
+            new PineconeConfig("");
+        } catch (PineconeConfigurationException expected) {
+            assertEquals(expected.getLocalizedMessage(), "The API key is required and must not be empty or null");
+        }
+    }
+
+    @Test
+    public void testGetUserAgent() {
+        PineconeConfig config = new PineconeConfig("testApiKey");
+        assertEquals(config.getUserAgent(), "lang=java; pineconeClientVersion=v0.8.0");
+    }
+
+    @Test
+    public void testGetUserAgentGrpc() {
+        PineconeConfig config = new PineconeConfig("testApiKey");
+        assertEquals(config.getUserAgentGrpc(), "lang=java; pineconeClientVersion[grpc]=v0.8.0");
+    }
+    @Test
+    public void testGetUserAgentWithSourceTag() {
+        PineconeConfig config = new PineconeConfig("testApiKey");
+        config.setSourceTag("testSourceTag");
+        assertEquals(config.getUserAgent(), "lang=java; pineconeClientVersion=v0.8.0; source_tag=testsourcetag");
+    }
+
+    @Test
+    public void testGetUserAgentGrpcWithSourceTag() {
+        PineconeConfig config = new PineconeConfig("testApiKey");
+        config.setSourceTag("testSourceTag");
+        assertEquals(config.getUserAgentGrpc(), "lang=java; pineconeClientVersion[grpc]=v0.8.0; source_tag=testsourcetag");
+    }
+
+    @Test
+    public void testSourceTagIsNormalized() {
+        PineconeConfig config = new PineconeConfig("testApiKey");
+        config.setSourceTag("test source tag !! @@ ##");
+        assertEquals(config.getSourceTag(), "test_source_tag___");
+
+        config.setSourceTag("TEST SOURCE Tag     ----");
+        assertEquals(config.getSourceTag(), "test_source_tag_");
```

Comment:
So I think if someone gives us a source tag of `test source tag !! @@ ##` we should turn that into: `test_source_tag`

Reply:
And I realized I needed to make an update to the Python impl of this. Here is what it looks like now. Note the inclusion of the `' '` space char to the charset regex, and then later the spaces will get replaced with underscore

```python
def _build_source_tag_field(source_tag):
    # normalize source tag
    # 1. Limit charset to [a-z0-9_ ]
    # 2. Lowercase
    # 3. Trim left/right whitespace
    # 4. Condense multiple spaces to one, and replace with underscore
    tag = re.sub(r'[^a-z0-9_ ]', '', source_tag)
    tag = tag.lower().strip()
    tag = "_".join(tag.split())
    return f"{SOURCE_TAG}={tag}"
```

Reply:
Agreed, and thanks for the catch. I've updated the Java implementation to match.